File: src/main/java/lzhang/leetcode/Six.java
package lzhang.leetcode;

import java.lang.reflect.Array;
import java.util.LinkedList;
import java.util.Queue;

/*The string "PAYPALISHIRING" is written in a zigzag pattern on a given number of 
 * rows like this: (you may want to display this pattern in a fixed font 
 * for better legibility)
 * P   A   H   N
 * A P L S I I G
 * Y   I   R
 *
 *And then read line by line: "PAHNAPLSIIGYIR"
 */
public class Six {
  @SuppressWarnings("unchecked")
  public String convert(String s, int numRows) {
    if (numRows>=s.length() || numRows==1){
      return s;
    }
    Queue<Character> tmp=new LinkedList<>();
    Queue<Character>[] data=
        (Queue<Character>[]) Array.newInstance(tmp.getClass(), numRows);
//    for (Queue<Character> init:data){
//      init=new LinkedList<>();
//    }
    for (int i=0;i<numRows;i++){
      data[i]=new LinkedList<Character>();
    }
    boolean down=true;
    int row=0;
    for (int i=0;i<s.length();i++){
      data[row].add(s.charAt(i));
      if (down){
        if (row==numRows-1){
          down=false;
          row--;
        } else {
          row++;
        }
      } else {
        if (row==0){
          down=true;
          row++;
        } else { 
          row--;
        }
      }
    }
    String result="";
    for (int i=0;i<numRows;i++){
      while (!data[i].isEmpty()){
        result+=data[i].poll();
      }
    }
    return result;
  }
}

File: javascript/six.js
/**
 * @param {string} s
 * @param {number} numRows
 * @return {string}
 */

var convert = function(s, numRows) {
	var map, numCols;
	if (numRows===1){
		return s;
	}
	var result='';
	map=[];
	for (var i=0; i<numRows; i++){
		map[i]=[];
	}
	var down=true;
	numCols=0;
	var rol=0;
	for (var i=0; i<s.length; i++){
		map[rol][numCols]=s[i];
		if (down){
			if (rol===numRows-1){
				down=false;
				rol--;
				numCols++;
			} else {
				rol++;				
			}
		} else {
			if (rol===0){
				down=true;
				rol++;
				numCols++;
			} else {
				rol--;
			}
		}
	}

	for (var i=0; i<numRows; i++){
		for (var j=0; j<=numCols; j++){
			if (map[i][j]!==undefined){
				result+=map[i][j];
			}
		}
	}

	return result;
};

console.log(convert("PAYPALISHIRING", 3));
